flash_io.c

```diff
@@ -0,0 +1,197 @@
+#include <hardware/flash.h>
+#include <hardware/sync.h>
+
+#include <stdio.h>
+#include <string.h>
+
+/*
+ *   The main purpose of saving user data to flash memory is to
+ *   handle unexpected loss of electrical power.
+ *   In such a case, the furnace driver can boot itself up
+ *   with previous config.
+ *
+ *   In order to read from flash,
+ *   config (both size and initialized targets) must be matching.
+ *
+ *   Also, we are reading from flash if and only if
+ *   the last record is the same as current config.
+ *   It is done in order to NOT boot with obsolete data.
+ *
+ *  memory layout:
+ *    size
+ *    targets
+ *    pilot.des_temp
+ *    pilot.is_enabled
+ *    log_bits
+ *    pwm_water
+ *
+ *    It is worth to note that you cannot rewrite flash memory,
+ *    once you write something in it,
+ *    you cannot overwrite this without erasing whole flash sector.
+ *    It is not described in pico-sdk.
+ *    It results from NOR/NAND flash memory.
+ */
+
+enum __attribute__((packed))
+init_targets {
+  MAGNETRON = 1 << 0,
+  WATER     = 1 << 1
+};
+
+// get symbols from the linker
+extern size_t
+FLASH_USER_DATA_START;
+
+extern size_t
+FLASH_USER_DATA_END;
+
+static const uint8_t
+FLASH_USER_DATA_LEN = sizeof(flash_io_t);
+
+static const void*
+flash_ptr;
+
+static const enum init_targets
+targets = 0
+  #if CONFIG_MAGNETRON
+    + MAGNETRON
+  #endif
+  #if CONFIG_WATER
+    + WATER
+  #endif
+;
+
+static bool
+check_end_of_user_flash()
+{
+  if(flash_ptr == &FLASH_USER_DATA_END)
```

Comment:
```diff
-    if(flash_ptr == &FLASH_USER_DATA_END)
+    if(flash_ptr >= &FLASH_USER_DATA_END)
```
?

Reply:
The logic with == is working fine. Although I think I know what you are trying to say.
Someone can come and change how flash_ptr was meant to work. I supposed that flash_ptr will never exceed &FLASH_USER_DATA_END, but someone can change and go out of our scope.
Because of that, I will change to >=, it is much safer for future development